File: backend/sdk/handlers/character_design_tools.py
"""
Character design tools for comprehensive character creation.

Provides tools for creating detailed characters with backstory and consolidated memory.
Used by detailed_character_designer agent during onboarding.
"""

import logging
from typing import Any

from services.agent_filesystem_service import AgentFilesystemService
from services.world_service import WorldService

from sdk.handlers.context import ToolContext
from sdk.loaders import is_tool_enabled
from sdk.tools.character_design import (
    CreateComprehensiveCharacterInput,
    ImplantConsolidatedMemoryInput,
)

logger = logging.getLogger("CharacterDesignTools")


def create_character_design_tools(ctx: ToolContext) -> list:
    """Create character design tools for comprehensive character creation.

    Args:
        ctx: Unified tool context containing agent info and dependencies

    Returns:
        List of character design tool functions
    """
    import crud
    from services.agent_factory import AgentFactory
    from services.room_mapping_service import RoomMappingService

    from sdk.handlers.common import build_action_context

    tools = []

    # Extract context variables
    world_name = ctx.world_name
    world_id = ctx.world_id
    room_id = ctx.room_id
    db = ctx.db

    # Import tool definitions
    if is_tool_enabled("create_comprehensive_character"):

        async def create_comprehensive_character_tool(args: dict[str, Any]):
            """Create a comprehensive character with detailed backstory.

            Used by detailed_character_designer during onboarding to create
            rich, memorable characters with depth and complexity.
            """
            validated = CreateComprehensiveCharacterInput(**args)

            logger.info(
                f"create_comprehensive_character: {validated.name} ({validated.role}) in {world_name}"
            )

            try:
                # Create agent in filesystem
                agent_name = validated.name.replace(" ", "_")

                # Build comprehensive agent files
                # in_a_nutshell: Brief identity (name + role)
                in_a_nutshell = f"{validated.name} is a {validated.role}."

                # characteristics: Full personality, appearance, backstory
                characteristics = f"""## Role
{validated.role}

## Appearance
{validated.appearance}

## Personality
{validated.personality}

## Backstory
{validated.backstory}

## Initial Disposition
{validated.initial_disposition}"""

                if validated.secret:
                    characteristics += f"\n\n## Hidden Detail\n{validated.secret}"

                # Create agent folder and files
                AgentFilesystemService.create_agent(
                    world_name, agent_name, in_a_nutshell, characteristics
                )

                # Determine target location
                which_location = validated.which_location or "current"

                # Get current location info
                context = build_action_context(world_name, "character creation")
                current_location_name = context.current_location

                # Update filesystem state for location
                if which_location.lower() != "current":
                    target_loc = await crud.get_location_by_name(db, world_id, which_location)
                    if target_loc:
                        room_key = RoomMappingService.location_to_room_key(target_loc.name)
                        if target_loc.room_id:
                            RoomMappingService.ensure_room_mapping_exists(
                                world_name=world_name,
                                room_key=room_key,
                                db_room_id=target_loc.room_id,
                                agents=[],
                            )
                        RoomMappingService.add_agent_to_room(world_name, room_key, agent_name)
                        location_display = target_loc.name
                    else:
                        # Fallback to current location
                        room_key = RoomMappingService.location_to_room_key(current_location_name)
                        RoomMappingService.add_agent_to_room(world_name, room_key, agent_name)
                        location_display = "current location"
                else:
                    room_key = RoomMappingService.location_to_room_key(current_location_name)
                    RoomMappingService.add_agent_to_room(world_name, room_key, agent_name)
                    location_display = "current location"

                # Create agent in DB
                new_agent = await AgentFactory.create_from_config(
                    db=db,
                    name=agent_name,
                    config_file=f"worlds/{world_name}/agents/{agent_name}",
                    group=None,
                )

                # Add to room/location in DB
                if which_location.lower() == "current":
                    # Check if this is the onboarding room (not a location room)
                    world = await crud.get_world(db, world_id)
                    is_onboarding_room = world and world.onboarding_room_id == room_id

                    if not is_onboarding_room:
                        # Only add to room if it's a location room (active gameplay)
                        from crud.room_agents import add_agent_to_room

                        await add_agent_to_room(db, room_id, new_agent.id)
                    else:
                        logger.info(
                            f"Skipping room addition during onboarding - {agent_name} will be placed at initial location"
                        )
                else:
                    target_loc = await crud.get_location_by_name(db, world_id, which_location)
                    if target_loc:
                        await crud.add_character_to_location(db, new_agent.id, target_loc.id)

                logger.info(f"Created comprehensive character: {validated.name} at {location_display}")

                # Build response
                memory_note = ""
                if validated.initial_memories:
                    memory_note = f"\n\n**Note:** Call `implant_consolidated_memory` next to add {len(validated.initial_memories)} initial memories."

                response_text = f"""**Comprehensive Character Created:**

**Basic Info:**
- Name: {validated.name}
- Role: {validated.role}
- Location: {location_display}
- Disposition: {validated.initial_disposition}

**Profile:**
- Appearance: {len(validated.appearance)} chars
- Personality: {len(validated.personality)} chars
- Backstory: {len(validated.backstory)} chars
- Secret: {"Yes" if validated.secret else "None"}

**Files Created:**
- `in_a_nutshell.md` - Brief identity
- `characteristics.md` - Full profile with backstory{memory_note}"""

                return {"content": [{"type": "text", "text": response_text}]}

            except Exception as e:
                logger.error(f"create_comprehensive_character error: {e}", exc_info=True)
                return {
                    "content": [{"type": "text", "text": f"Error creating comprehensive character: {e}"}],
                    "is_error": True,
                }

        tools.append(create_comprehensive_character_tool)

    if is_tool_enabled("implant_consolidated_memory"):

        async def implant_consolidated_memory_tool(args: dict[str, Any]):
            """Implant consolidated memories into a character's memory file.

            Populates consolidated_memory.md with formatted memories.
            """
            validated = ImplantConsolidatedMemoryInput(**args)

            logger.info(
                f"implant_consolidated_memory: {validated.character_name} - {len(validated.memories)} memories ({validated.mode})"
            )

            try:
                # Convert character name to filesystem-safe name
                agent_name = validated.character_name.replace(" ", "_")

                # Get agent path
                world_path = WorldService.get_world_path(world_name)
                agent_path = world_path / "agents" / agent_name

                if not agent_path.exists():
                    return {
                        "content": [
                            {
                                "type": "text",
                                "text": f"Error: Character '{validated.character_name}' not found in world. "
                                f"Create the character first using create_comprehensive_character.",
                            }
                        ],
                        "is_error": True,
                    }

                # Prepare consolidated_memory.md path
                memory_file = agent_path / "consolidated_memory.md"

                # Build memory content in correct format
                memory_content_parts = []

                # If append mode and file exists, read existing content
                if validated.mode == "append" and memory_file.exists():
                    existing_content = memory_file.read_text(encoding="utf-8").strip()
                    if existing_content:
                        memory_content_parts.append(existing_content)

                # Add new memories
                for memory in validated.memories:
                    memory_section = f"## [{memory.subtitle}]\n{memory.content}"
                    memory_content_parts.append(memory_section)

                # Join all parts with double newline
                final_content = "\n\n".join(memory_content_parts)

                # Write to file
                memory_file.write_text(final_content, encoding="utf-8")

                logger.info(
                    f"Implanted {len(validated.memories)} memories for {validated.character_name} ({validated.mode} mode)"
                )

                # Build response
                operation_text = "added" if validated.mode == "append" else "set"
                total_memories = len(validated.memories)
                if validated.mode == "append" and memory_file.exists():
                    # Count total ## headers in file
                    total_memories = final_content.count("## [")

                response_text = f"""**Consolidated Memories Implanted:**

**Character:** {validated.character_name}
**Operation:** {validated.mode}
**Memories {operation_text}:** {len(validated.memories)}
**Total memories in file:** {total_memories}

**Memory Subtitles:**
"""
                for i, memory in enumerate(validated.memories, 1):
                    response_text += f"{i}. [{memory.subtitle}]\n"

                response_text += """
**File:** `consolidated_memory.md`
**Format:** Memories stored with `## [subtitle]` headers

The character can now recall these memories using the `recall` tool."""

                return {"content": [{"type": "text", "text": response_text}]}

            except Exception as e:
                logger.error(f"implant_consolidated_memory error: {e}", exc_info=True)
                return {
                    "content": [{"type": "text", "text": f"Error implanting memories: {e}"}],
                    "is_error": True,
                }

        tools.append(implant_consolidated_memory_tool)

    return tools

File: backend/src/sdk/handlers/character-design-tools.ts
import { existsSync, readFileSync, writeFileSync } from 'node:fs'
import { join } from 'node:path'
import { addCharacterToLocation, getLocationByName } from '../../crud/locations'
import { addAgentToRoom } from '../../crud/rooms'
import { getWorld } from '../../crud/worlds'
import type { PlayerService } from '../../services/player-service'
import { RoomMappingService } from '../../services/room-mapping'
import {
  createComprehensiveCharacterTool,
  implantConsolidatedMemoryTool,
} from '../tools/character-design'
import { getLogger } from '../../infrastructure/logging/logger'
import type { AgentFactory } from '../../services/agent-factory'
import type { AgentFilesystemService } from '../../services/agent-filesystem-service'
import type { WorldService } from '../../services/world-service'
import { tool, requireWorldId, requireWorldName, toolError, toolSuccess, type SdkTool, type ToolContext } from './context'

/**
 * Deep character creation for onboarding. These two are a pair, called in order:
 * the first writes the character's files and places them, the second fills
 * `consolidated_memory.md`, which is what `recall` reads. Skipping the second
 * leaves an NPC who cannot remember their own history.
 *
 * **These tools cannot be overridden or disabled from a group config** — they go
 * through no registry lookup. See `tools/registry.ts::CHARACTER_DESIGN_GROUP`.
 */

const logger = getLogger('CharacterDesignTools')

export interface CharacterDesignDeps {
  agentFiles: AgentFilesystemService
  agentFactory: AgentFactory
  players: PlayerService
  rooms: RoomMappingService
  /** Only for resolving the agent folder path — `agentPath` is service-private. */
  worlds: WorldService
}

export function createCharacterDesignTools(
  ctx: ToolContext,
  deps: CharacterDesignDeps,
): SdkTool[] {
  const worldName = requireWorldName(ctx)
  const worldId = requireWorldId(ctx)

  const createCharacter = tool(
    createComprehensiveCharacterTool.name,
    createComprehensiveCharacterTool.description,
    createComprehensiveCharacterTool.inputSchema,
    async (args) => {
      try {
        const db = ctx.getDb()
        const agentName = args.name.replaceAll(' ', '_')

        const inANutshell = `${args.name} is a ${args.role}.`
        let characteristics = `## Role\n${args.role}\n\n## Appearance\n${args.appearance}\n\n## Personality\n${args.personality}\n\n## Backstory\n${args.backstory}\n\n## Initial Disposition\n${args.initial_disposition}`
        if (args.secret) characteristics += `\n\n## Hidden Detail\n${args.secret}`

        deps.agentFiles.createAgent(worldName, agentName, inANutshell, characteristics)

        const whichLocation = args.which_location || 'current'
        const isCurrent = whichLocation.toLowerCase() === 'current'
        const currentLocation = deps.players.loadPlayerState(worldName)?.currentLocation || 'unknown'

        let locationDisplay = 'current location'
        let targetLocation = isCurrent ? null : getLocationByName(db, worldId, whichLocation)

        if (!isCurrent && targetLocation) {
          const roomKey = RoomMappingService.locationToRoomKey(targetLocation.name)
          if (targetLocation.roomId) {
            deps.rooms.ensureRoomMappingExists(worldName, roomKey, targetLocation.roomId, [])
          }
          deps.rooms.addAgentToRoom(worldName, roomKey, agentName)
          // The *folder* name, not `display_name`: the one place the two
          // character-creation tools disagree.
          locationDisplay = targetLocation.name
        } else {
          targetLocation = null
          deps.rooms.addAgentToRoom(
            worldName,
            RoomMappingService.locationToRoomKey(currentLocation),
            agentName,
          )
        }

        const newAgent = deps.agentFactory.createFromConfig(db, {
          name: agentName,
          configFile: `worlds/${worldName}/agents/${agentName}`,
          group: null,
        })

        if (isCurrent) {
          const world = getWorld(db, worldId)
          const isOnboardingRoom =
            world !== null && ctx.roomId !== undefined && world.onboardingRoomId === ctx.roomId
          if (!isOnboardingRoom && ctx.roomId !== undefined) {
            addAgentToRoom(db, ctx.roomId, newAgent.id)
          }
        } else if (targetLocation) {
          addCharacterToLocation(db, newAgent.id, targetLocation.id)
        }

        const memoryNote =
          args.initial_memories && args.initial_memories.length > 0
            ? `\n\n**Note:** Call \`implant_consolidated_memory\` next to add ${args.initial_memories.length} initial memories.`
            : ''

        return toolSuccess(
          `**Comprehensive Character Created:**\n\n` +
            `**Basic Info:**\n` +
            `- Name: ${args.name}\n` +
            `- Role: ${args.role}\n` +
            `- Location: ${locationDisplay}\n` +
            `- Disposition: ${args.initial_disposition}\n\n` +
            `**Profile:**\n` +
            `- Appearance: ${args.appearance.length} chars\n` +
            `- Personality: ${args.personality.length} chars\n` +
            `- Backstory: ${args.backstory.length} chars\n` +
            `- Secret: ${args.secret ? 'Yes' : 'None'}\n\n` +
            `**Files Created:**\n` +
            `- \`in_a_nutshell.md\` - Brief identity\n` +
            `- \`characteristics.md\` - Full profile with backstory${memoryNote}`,
        )
      } catch (error) {
        logger.error(`create_comprehensive_character error: ${String(error)}`)
        return toolError(`Error creating comprehensive character: ${String(error)}`)
      }
    },
  )

  const implantMemory = tool(
    implantConsolidatedMemoryTool.name,
    implantConsolidatedMemoryTool.description,
    implantConsolidatedMemoryTool.inputSchema,
    async (args) => {
      try {
        const agentName = args.character_name.replaceAll(' ', '_')
        const agentPath = join(deps.worlds.getWorldPath(worldName), 'agents', agentName)

        if (!existsSync(agentPath)) {
          return toolError(
            `Error: Character '${args.character_name}' not found in world. ` +
              `Create the character first using create_comprehensive_character.`,
          )
        }

        const memoryFile = join(agentPath, 'consolidated_memory.md')
        const fileExisted = existsSync(memoryFile)

        const parts: string[] = []
        if (args.mode === 'append' && fileExisted) {
          const existing = readFileSync(memoryFile, 'utf-8').trim()
          if (existing) parts.push(existing)
        }
        for (const memory of args.memories) {
          // `## [subtitle]` exactly: `memory.ts` keys the long-term memory index
          // off the bracketed form, so an unbracketed heading is invisible to
          // `recall`.
          parts.push(`## [${memory.subtitle}]\n${memory.content}`)
        }

        const finalContent = parts.join('\n\n')
        writeFileSync(memoryFile, finalContent, 'utf-8')

        const operationText = args.mode === 'append' ? 'added' : 'set'
        // Either branch yields the total in the file, which is what the label
        // claims.
        const totalMemories =
          args.mode === 'append' && fileExisted
            ? (finalContent.match(/## \[/g)?.length ?? 0)
            : args.memories.length

        let text =
          `**Consolidated Memories Implanted:**\n\n` +
          `**Character:** ${args.character_name}\n` +
          `**Operation:** ${args.mode}\n` +
          `**Memories ${operationText}:** ${args.memories.length}\n` +
          `**Total memories in file:** ${totalMemories}\n\n` +
          `**Memory Subtitles:**\n`
        args.memories.forEach((memory, index) => {
          text += `${index + 1}. [${memory.subtitle}]\n`
        })
        text +=
          `\n**File:** \`consolidated_memory.md\`\n` +
          `**Format:** Memories stored with \`## [subtitle]\` headers\n\n` +
          `The character can now recall these memories using the \`recall\` tool.`

        return toolSuccess(text)
      } catch (error) {
        logger.error(`implant_consolidated_memory error: ${String(error)}`)
        return toolError(`Error implanting memories: ${String(error)}`)
      }
    },
  )

  return [createCharacter, implantMemory]
}
